### PR TITLE
[Feature] Add ENCNet and SECrossEntropyLoss

### PR DESCRIPTION
**Pull request by justld**
说明：
添加了ENCNet和Semantic Encoding Loss
configs/encnet/encnet_resnet101_os8_cityscapes_1024_512_80k.yml
configs/encnet/README.md
paddleseg/models/encnet.py
paddleseg/models/init.py
paddleseg/models/losses/init.py
paddleseg/models/losses/semantic_encode_cross_entropy_loss.py

### CODE REVIEW

File: paddleseg/models/losses/semantic_encode_cross_entropy_loss.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        ignore_index (int): Specifies a target value that is ignored
+            and does not contribute to the input gradient. Default: 255.
+    """
+    def __init__(self, num_classes, ignore_index=255):
+        super(SECrossEntropyLoss, self).__init__()
+        self.num_classes = num_classes
+
+    def forward(self, logit, label):
+        if logit.ndim == 4:
+            logit_ = logit.sequeeze(2).sequeeze(3)
```
Comment (shiyutang) [inline paddleseg/models/losses/semantic_encode_cross_entropy_loss.py]: Is "sequeeze" actually squeeze?

File: paddleseg/models/encnet.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+                    ))
+            self.fusion = layers.ConvBNReLU(
+                len(in_channels) * mid_channels,
+                mid_channels,
+                3,
+                padding=1,
+            )
+
+        self.enc_module = EncModule(mid_channels, num_codes)
+        self.head = nn.Sequential(nn.Conv2D(mid_channels, num_classes, 1), )
```
Comment (shiyutang) [inline paddleseg/models/encnet.py]: It seems nn.Sequential is not necessary

File: paddleseg/models/encnet.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+    def forward(self, inputs):
+        N, C, H, W = paddle.shape(inputs)
+        feats = self.backbone(inputs)
+        fcn_feat = feats[2]
+
+        temp_feats = []
+        for i in self.backbone_indices:
+            temp_feats.append(feats[i])
+        feats = temp_feats
```
Comment (shiyutang) [inline paddleseg/models/encnet.py]: feats = [feats[i] for i in self.backbone_indices]

File: paddleseg/models/encnet.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+            Encoding(channels=in_channels, num_codes=num_codes),
+            nn.BatchNorm1D(num_codes),
+            nn.ReLU(),
+        )
+        self.fc = nn.Sequential(
+            nn.Linear(in_channels, in_channels),
+            nn.Sigmoid(),
+        )
+        self.in_channels = in_channels
+        self.num_codes = num_codes
```
Comment (shiyutang) [inline paddleseg/models/encnet.py]: self.num_codes is not used again

File: paddleseg/models/encnet.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+    def forward(self, x):
+        encoding_projection = self.encoding_project(x)
+        encoding_feat = self.encoding(encoding_projection)
+
+        encoding_feat = encoding_feat.mean(axis=1)
+        batch_size, channels, _, _ = paddle.shape(x)
```
Comment (shiyutang) [inline paddleseg/models/encnet.py]: channels is not used

File: paddleseg/models/losses/semantic_encode_cross_entropy_loss.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    The Semantic Encoding Loss implementation based on PaddlePaddle.
+
+    Args:
+        num_classes (int): The unique number of target classes.
+        ignore_index (int): Specifies a target value that is ignored
+            and does not contribute to the input gradient. Default: 255.
+    """
+    def __init__(self, num_classes, ignore_index=255):
+        super(SECrossEntropyLoss, self).__init__()
+        self.num_classes = num_classes
```
Comment (shiyutang) [inline paddleseg/models/losses/semantic_encode_cross_entropy_loss.py]: why do we need to assign num_class by the user?  it equals to logit.shape[1] which can be read by us.

**Review by shiyutang — CHANGES_REQUESTED** [review]

File: paddleseg/models/losses/semantic_encode_cross_entropy_loss.py (hunk 1 above)
Comment (justld) [reply paddleseg/models/losses/semantic_encode_cross_entropy_loss.py]: fixed

File: paddleseg/models/encnet.py (hunk 2 above)
Comment (justld) [reply paddleseg/models/encnet.py]: removed

File: paddleseg/models/encnet.py (hunk 4 above)
Comment (justld) [reply paddleseg/models/encnet.py]: deleted

File: paddleseg/models/losses/semantic_encode_cross_entropy_loss.py (hunk 6 above)
Comment (justld) [reply paddleseg/models/losses/semantic_encode_cross_entropy_loss.py]: modified

File: paddleseg/models/encnet.py (hunk 3 above)
Comment (justld) [reply paddleseg/models/encnet.py]: fixed

File: paddleseg/models/encnet.py (hunk 5 above)
Comment (justld) [reply paddleseg/models/encnet.py]: deleted

**Review by shiyutang — DISMISSED** [review]
LGTM

**Comment by shiyutang** [comment]
This PR need to retrain before merge

**Review by shiyutang — APPROVED** [review]

**Review by michaelowenliu — APPROVED** [review]
